core/src/Kokkos_View.hpp

```diff
@@ -792,33 +792,54 @@ class View : public ViewTraits<DataType, Properties...> {
   template <class Space, bool = Kokkos::Impl::MemorySpaceAccess<
                              Space, typename traits::memory_space>::accessible>
   struct verify_space {
-    KOKKOS_FORCEINLINE_FUNCTION static void check() {}
+    template <typename... Args>
+    KOKKOS_FORCEINLINE_FUNCTION static void check(Args&&...) {}
   };
 
   template <class Space>
   struct verify_space<Space, false> {
-    KOKKOS_FORCEINLINE_FUNCTION static void check() {
+    template <typename Tracker, typename MapType, typename... IdxType>
+    KOKKOS_FORCEINLINE_FUNCTION static void check(const Tracker&
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+                                                      tracker
+#endif
+                                                  ,
+                                                  const MapType&,
+                                                  const IdxType...) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      std::string error_msg =
+          "Kokkos::View ERROR: attempt to access View " +
+          std::string(
+              tracker.m_tracker
+                  .template get_label<typename traits::memory_space>()) +
+          " in memory space " +
+          std::string(typename traits::memory_space{}.name()) +
+          " from execution space " + std::string(Space{}.name());
+      Kokkos::abort(error_msg.c_str());
+#else
       Kokkos::abort(
-          "Kokkos::View ERROR: attempt to access inaccessible memory space");
-    };
+          "Kokkos::View ERROR: attempt to access a View from the device in an "
+          "inappropriate space.");
+#endif
+    }
   };
 
 #if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
 
 #define KOKKOS_IMPL_SINK(ARG) ARG
 
-#define KOKKOS_IMPL_VIEW_OPERATOR_VERIFY(ARG)             \
-  View::template verify_space<                            \
-      Kokkos::Impl::ActiveExecutionMemorySpace>::check(); \
+#define KOKKOS_IMPL_VIEW_OPERATOR_VERIFY(ARG)                                  \
+  View::template verify_space<Kokkos::Impl::ActiveExecutionMemorySpace>::check \
+      ARG;                                                                     \
   Kokkos::Impl::view_verify_operator_bounds<typename traits::memory_space> ARG;
 
 #else
 
-#define KOKKOS_IMPL_SINK(ARG)
+#define KOKKOS_IMPL_SINK(ARG) ARG
```

Comment:
```suggestion
#define KOKKOS_IMPL_SINK(ARG)
```
Do we need to change this?

Reply:
When I didn't have that I had compiler errors, but I have no idea what this forest of preprocessor stuff does. If you can figure out how to make it work without this I'm game

Reply:
Yes, we now always use `ARG` so we don't need `KOKKOS_IMPL_SINK` anymore. I removed it.